main.php

```diff
@@ -1,226 +1,19 @@
-			return;
-		}
-
-		// Fail and redirect the user to the login page
-		$loginPage = Director::absoluteURL(Security::config()->login_url);
-		$loginPage .= "?BackURL=" . urlencode($_SERVER['REQUEST_URI']);
-		header('location: '.$loginPage, true, 302);
-		die;
-	})
-	// Finally if a token was requested but there was an error while figuring out if it's allowed, do it anyway
-	->thenIfErrored(function() use ($reloadToken){
-		if ($reloadToken) {
-			$reloadToken->reloadWithToken();
-		}
-	})
-	->execute();
-
-global $databaseConfig;
-
-// Redirect to the installer if no database is selected
-if(!isset($databaseConfig) || !isset($databaseConfig['database']) || !$databaseConfig['database']) {
-
-    // Is there an _ss_environment.php file?
-    if(file_exists(BASE_PATH . '/_ss_environment.php') || file_exists(dirname(BASE_PATH) . '/_ss_environment.php')) {
-        header($_SERVER['SERVER_PROTOCOL'] . " 500 Server Error");
-        $dv = new SilverStripe\Dev\DebugView();
-        echo $dv->renderHeader();
-        echo $dv->renderInfo(
-            "Configuraton Error",
-            Director::absoluteBaseURL()
-        );
-        echo $dv->renderParagraph(
-            'You need to replace your _ss_environment.php file with a .env file, or with environment variables.<br><br>'
-            . 'See the <a href="https://docs.silverstripe.org/en/4/getting_started/environment_management/">'
-            . 'Environment Management</a> docs for more information.'
-        );
-        echo $dv->renderFooter();
-
-        die();
-    }
-
-	if(!file_exists(BASE_PATH . '/install.php')) {
-		header($_SERVER['SERVER_PROTOCOL'] . " 500 Server Error");
-		die('SilverStripe Framework requires a $databaseConfig defined.');
-	}
-	$host = isset($_SERVER['HTTP_HOST']) ? $_SERVER['HTTP_HOST'] : $_SERVER['SERVER_NAME'];
-	$s = (isset($_SERVER['SSL']) || (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] != 'off')) ? 's' : '';
-	$installURL = "http$s://" . $host . BASE_URL . '/install.php';
-
-	// The above dirname() will equate to "\" on Windows when installing directly from http://localhost (not using
-	// a sub-directory), this really messes things up in some browsers. Let's get rid of the backslashes
-	$installURL = str_replace('\\', '', $installURL);
-
-	header("Location: $installURL");
-	die();
-}
-
-// Direct away - this is the "main" function, that hands control to the appropriate controller
-DataModel::set_inst(new DataModel());
-Director::direct($url, DataModel::inst());
+use SilverStripe\Core\AppKernel;
+use SilverStripe\Core\HTTPApplication;
+use SilverStripe\Core\Startup\ErrorControlChainMiddleware;
+use SilverStripe\Core\Startup\OutputMiddleware;
+use SilverStripe\Control\HTTPRequest;
+
+require __DIR__ . '/src/includes/autoload.php';
+
+// Build request and detect flush
+$request = HTTPRequest::createFromEnvironment();
+
+// Default application
+$kernel = new AppKernel();
+$app = new HTTPApplication($kernel);
+$app->addMiddleware(new OutputMiddleware());
```

Comment:
My response to ingo is that this would weaken the ability of middleware to control response handling. E.g. what if a middleware needed to perform some action after ->output()?